pysrc/bytewax/execution.py

```diff
@@ -189,26 +411,85 @@ def run_cluster(
+            communicating to subprocesses. Defaults to total number of
+            worker threads in cluster * 2.
 
-        List of `(epoch, item)` tuples seen by capture operators.
+    Yields:
+
+        `(epoch, item)` tuples seen by capture operators.
 
     """
-    man = Manager()
-    inp = man.list(list(inp))
+    worker_count = proc_count * worker_count_per_proc
+    if ipc_buffer_size is None:
+        ipc_buffer_size = worker_count * 2
+    manager = Manager()
+
+    in_q = manager.Queue(ipc_buffer_size)
 
     def input_builder(worker_index, worker_count):
-        for i, epoch_item in enumerate(inp):
-            if i % worker_count == worker_index:
-                yield epoch_item
+        def in_get():
+            while True:
+                msg, data = in_q.get()
+                if msg == "yield":
+                    yield data
+                elif msg == "done":
+                    return
 
-    out = man.list()
+        return in_get()
+
+    out_q = manager.Queue(ipc_buffer_size)
 
     def output_builder(worker_index, worker_count):
-        return out.append
+        def out_put(epoch_item):
+            out_q.put(epoch_item)
+
+        return out_put
 
-    spawn_cluster(
+    inp_iter = iter(inp)
+    coro = spawn_cluster(
         flow, input_builder, output_builder, proc_count, worker_count_per_proc
     )
 
-    return out
+    work_remains = True
+    input_remains = True
+    # Use an input buffer in this proc to store items we've
+    # irreversibly popped off the input iterator in case the worker
+    # input queue is full.
+    input_buffer = deque()
+    while work_remains or input_remains or len(input_buffer) > 0 or not out_q.empty():
+        # 1. Check on workers.
+        if work_remains:
+            work_remains = _step_coro(coro)
+            # If input / output queues are full / empty, this'll be a
+            # hot busy loop so slow down to wait for IO to do.
+            time.sleep(0.001)
+
+        # 2. Read and buffer some input.
+        while input_remains and len(input_buffer) < ipc_buffer_size:
+            try:
+                msg = ("yield", next(inp_iter))
+                input_buffer.appendleft(msg)
+            except StopIteration:
+                input_remains = False
+                # If the input is complete, enqueue messages to
+                # workers to stop work.
+                msg = ("done", None)
+                for _ in range(worker_count):
+                    input_buffer.appendleft(msg)
+
+        # 3. Send as much input as we can to workers.
+        while len(input_buffer) > 0:
+            try:
+                msg = input_buffer.pop()
+                in_q.put(msg, False)
+            except Full:
+                input_buffer.append(msg)
+                break
+
+        # 4. Get as much output as we can from workers.
+        while not out_q.empty():
+            try:
+                yield out_q.get(False)
```

Comment:
There is a race condition here because `out_q.empty` is not guaranteed to be correct. So the cluster might push final values into the `out_q`, shutdown, but then the output won't be yielded. We'll need to properly signal shutdown through the queue.